man/illegal2vrml.1
.\" See section COPYING for conditions for redistribution
.\"
.TH illegal2vrml 1 2017-09-20 "white_dune 0.99rc766"
.SH NAME
\fBillegal2vrml\fP \- convert VRML97 files with illegal extensions
.SH SYNOPSIS
.br
\fBillegal2vrml\fP
[\fB
-prefix prefix
\fP] 
\fB
protofile.wrl
\fP
\fB
file.wrl
\fP
.P
.SH DESCRIPTION
.B illegal2vrml
convert Virtual Reality Modeling Language (VRML97) files with illegal 
extensions to valid VRML97, according to ISO/IEC 14772 Annex F.
Annex F decribe, how extensions to the VRML language should be
implemented: using a EXTERNPROTO with a "urn" (fake URL) that sign the
VRML browser the extension and not to use the following URL.
.P
.B illegal2vrml
read the file 
.B protofile.wrl
which should contain such EXTERNPROTOs. 
The names of the EXTERNPROTOs should start with the same prefix.
The prefix is needed to deal with additional illegal field names
in valid VRML97 nodes.
.br
In the next step
.B illegal2vrml
read the file 
.B file.wrl
and replace all node names which are identical to the EXTERNPROTOs
without the prefix with the names of the EXTERNPROTOs (with the prefix).
.br
In the last step, the result is written to standard output.
.P
.SH OPTIONS
.TP
.B \-prefix prefix \fP 

set the name of the prefix of the EXTERNPROTOs. If this option is not
used, the prefix is calculated from the proto names in 
.B protofile.wrl
.P
.SH EXAMPLE
.P
Imagine a VRML97 file with a illegal extension to the WorldInfo node
.LP
.TP
$ cat $HOME/worldinfo.wrl
.br
#VRML V2.0 utf8

WorldInfo 
.br
{ 
.br
  copyright "(C) illegal extension Inc."
.br
  info      "a existing VRML97 node with a illegal extension"
.br
}
.LP
.TP
A matching file with a EXTERNPROTO would be
.LP
.TP
$ cat $HOME/protofile.wrl
.br
#VRML V2.0 utf8

EXTERNPROTO PrefixWorldInfo
.br
[
.br
field SFString copyright
.br
field MFString info
.br
field SFString title
.br
]
.br
["urn:inet:browser.com:library:PrefixWorldInfo" "PrefixWorldInfo.wrl"]
.LP
.TP
To convert to a valid VRML97 file, use
.LP
.TP
$ illegal2vrml -prefix Prefix  $HOME/protofile.wrl $HOME/worldinfo.wrl
.br
#VRML V2.0 utf8

EXTERNPROTO PrefixWorldInfo
.br
[
.br
field SFString copyright
.br
field MFString info
.br
field SFString title
.br
]
.br
["urn:inet:browser.com:library:PrefixWorldInfo" "PrefixWorldInfo.wrl"]

PrefixWorldInfo 
.br
  {
.br
  copyright "(C) illegal extension Inc."
.br
  info 
.br
    [
.br
    "a existing VRML97 node with a illegal extension"
.br
    ]
.br
  }
.P
.SH COPYRIGHT
    illegal2vrml convert VRML97 files with illegal extensions 
.br
    Copyright (C) 2000-2003  J. "MUFTI" Scheurich, Stephen F. White

    This program is free software; you can redistribute it 
    and/or modify it under the terms of the 
    GNU General Public License 
    as published by the Free Software Foundation; either 
    version 2 of the License, or (at your option) any later 
    version.
.P
.SH BUGS
Inside PROTOs, the conversion program do not parse for node names (yet).
Therefore only a string based replacement is made inside PROTOs.
.SH DIAGNOSTICS
Exit status is 0 for sucessfull operation,
1 if inputfiles can not be sucessfully read or other error 
.SH "SEE ALSO"
.TP
VRML97 International Standard
.br
http://www.web3d.org/x3d/specifications/vrml/ISO-IEC-14772-IS-VRML97WithAmendment1
.TP
VRML97 International Standard Annex F
.br
http://www.web3d.org/x3d/specifications/vrml/ISO-IEC-14772-IS-VRML97WithAmendment1/part1/extensions.html


src/illegal2vrml.cpp
/*
 * illegal2vrml.cpp
 *
 * Copyright (C) 1999 Stephen F. White, 2003 J. "MUFTI" Scheurich
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program (see the file "COPYING" for details); if 
 * not, write to the Free Software Foundation, Inc., 675 Mass Ave, 
 * Cambridge, MA 02139, USA.
 */

#include <stdio.h>
#include "stdafx.h"

#include "DuneApp.h"
#include "Scene.h"
#include "xerrorhandler.h"

#ifdef _WIN32
#  define errorprintf msgboxprintf
#else
#  define errorprintf fprintf   
#endif

static char* prefix=NULL;

static
bool illegal2vrmlParseCommandlineArgument(int & i,int argc, char** argv)
{
    bool found=false;
    if (strcmp(argv[i],"-prefix")==0) {
       if (i++>=argc) return found;
       found=true;
       prefix=argv[i];
#ifdef MACOSX
    } else if (stringncmp(argv[i],"-psn_")==0) {
       // parse special MacOSX commandline parameters
       found=true;
#endif
    } else
       found=false;
    return found;    
}

#define errormsg \
   "usage:  illegal2vrml [-prefix prefix] file_with_protos.wrl file.wrl\n"

int illegal2vrml(int argc, char *argv[])
{
    TheApp = new DuneApp();
    bool *files=new bool[argc+1];
    for (int i=2;i<argc;i++)
       files[i]=false;
    int firstFileIndex=0;
    int secondFileIndex=0;
    int i=2;
    while (i<argc) {
       files[i]=!illegal2vrmlParseCommandlineArgument(i,argc,argv);
       if (files[i]) {
          if (firstFileIndex==0)
             firstFileIndex=i;
          else if (secondFileIndex==0)
             secondFileIndex=i;
          else {
             errorprintf(stderr, errormsg);
             return 1;
          }
       }
       i++;
    }
    if ((i>argc) || (firstFileIndex==0)) {
       errorprintf(stderr, errormsg);
       return 1;
       }
    Scene *scene = new Scene();
    scene->setExternProtoWarning(false);
    if (!TheApp->AddFile(argv[firstFileIndex], scene)) {
       myperror(argv[firstFileIndex]);
       return(1);
    }
    if (!scene->setPrefix(prefix)) {
       errorprintf(stderr, "prefix can not be generated, use \"-prefix\"\n");
       errorprintf(stderr, errormsg);
       return(1);
    }
    if ((i>argc) || (secondFileIndex==0)) {
       errorprintf(stderr, errormsg);
       return 1;
       }
    scene->setExternProtoWarning(true);
    if (!TheApp->AddFile(argv[secondFileIndex], scene)) {
       myperror(argv[secondFileIndex]);
       return(1);
    }
    scene->write(1, "./export.wrl", TEMP_SAVE);
    delete TheApp;
    return 0;
}

